internal/api/query_freights_v1alpha1.go

```diff
@@ -0,0 +1,144 @@
+package api
+
+import (
+	"context"
+	"sort"
+
+	"connectrpc.com/connect"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kubev1alpha1 "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/api/types/v1alpha1"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+	apiv1alpha1 "github.com/akuity/kargo/pkg/api/v1alpha1"
+)
+
+const (
+	GroupByContainerRepository = "container_repo"
+	GroupByGitRepository       = "git_repo"
+	GroupByHelmRepository      = "helm_repo"
+
+	OrderByFirstSeen       = "first_seen"
+	OrderBySemanticVersion = "semantic_version"
+)
+
+func (s *server) QueryFreights(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.QueryFreightsRequest],
+) (*connect.Response[svcv1alpha1.QueryFreightsResponse], error) {
+	if req.Msg.GetProject() == "" {
+		return nil, connect.NewError(connect.CodeInvalidArgument, errors.New("project should not be empty"))
+	}
+	if err := s.validateProject(ctx, req.Msg.GetProject()); err != nil {
+		return nil, err
+	}
+	if err := validateGroupBy(req.Msg.GetGroupBy()); err != nil {
+		return nil, err
+	}
+	if err := validateOrderBy(req.Msg.GetOrderBy()); err != nil {
+		return nil, err
+	}
+
+	var stages []kubev1alpha1.Stage
+	if req.Msg.GetStage() != "" {
+		stage, err := getStage(ctx, s.client, req.Msg.GetProject(), req.Msg.GetStage())
+		if err != nil {
+			return nil, err
+		}
+		stages = append(stages, *stage)
+	} else {
+		var list kubev1alpha1.StageList
+		if err := s.client.List(ctx, &list, client.InNamespace(req.Msg.GetProject())); err != nil {
+			return nil, connect.NewError(connect.CodeInternal, err)
+		}
+		stages = list.Items
+	}
+
+	seen := make(map[string]kubev1alpha1.StageState)
```

Comment:
Nit: I know some people seem to be averse to the pattern, but you _could_ use set semantics here `seen := make(map[string]struct{})`

It's fewer allocations on the heap.

Reply:
Oh yes, my normal preference is actually:
```golang
make(map[string]bool)
```

So that the code is a bit more readable:
```golang
if seen[f.ID] {

}
```

which I prefer over:
```
if _, ok := seen[f.ID]; ok {

}
```

The reason it's a struct, is because at one point I thought I needed to do something with the actual object in the helper. But looking at the code, I ended up not doing anything with it. So I will change it to `make(map[string]bool)`